proofs/tree.go

```diff
@@ -12,20 +12,117 @@ Supported types:
 * timestamp.Timestamp
 
 
-Note: this is a basic implementation that lacks support for serializing more complex structs. The interfaces and
-functions in this library will change significantly in the near future.
-
-Advanced options
+Available Protobuf Options
 
 Fields can be excluded from the flattener by setting the custom protobuf option
 `proofs.exclude_from_tree` found in `proofs/proto/proof.proto`.
 
+Fields can be treated as raw (already hashed values) by setting the option `proofs.hashed_field`.
+
 	message Document {
 		string value_a = 1;
 		string value_b = 2 [
 			(proofs.exclude_from_tree) = true
 		];
+		bytes value_c = 3[
+			(proofs.hashed_field) = true
+		];
+	}
+
+Nested and Repeated Structures
+
+Nested and repeated fields will be flattened following a dotted notation. Given the following example:
+
+	message NestedDocument {
+	  string fieldA = 1;
+	  repeated Document fieldB = 2;
+	  repeated string fieldC = 3;
 	}
+
+	message Document {
+	  string fieldA = 1;
+	}
+
+	NestedDocument{
+		fieldA: "foobar",
+		fieldB: []Document{Document{fieldA: "1"}, Document{fieldA: "2"}},
+		fieldC: []string{"a", "b", "c"}
+	}
+
+A tree will be created out of this document by flattening all the fields values
+as leaves. This would result in a tree with the following leaves:
+
+ fieldA, fieldB.length, fieldB[0], fieldB[1], fieldC.length, fieldC[0], fieldC[1], fieldC[2]
+
+Proof format
+
+This library defines a proof format that ensures both human readable, concise and secure Merkle proofs:
+
+ {
+    "property":"ValueA",
+    "value":"Example",
+    "salt":"1VWQFUGCXl1AYS0iAULHQow4XEjgJF/TpAuOO2Rnm+E=",
+    "hashes":[
+        { "right":"kYXAGhDdPiFMq1ZQMOZiKmSf1S1eHNgJ6BIPSIExOj8=" },
```

Comment:
We are moving towards HEX fields, can you update this example godoc with a new sample?

Reply:
I don't think we should impose this on precise-proofs. This transformation for us happens at a later point but here I'd like to stick to standard protobuf.